settings-portal/Settings.vala

```diff
@@ -0,0 +1,156 @@
+}
+
+[DBus (name = "org.freedesktop.Accounts")]
+interface FDO.Accounts : Object {
+    public abstract string find_user_by_name (string username) throws GLib.Error;
+}
+
+private class AccountsServiceMonitor : GLib.Object {
+    private FDO.Accounts? accounts_service = null;
+    private Pantheon.AccountsService? pantheon_act = null;
+    private string user_path;
+
+    public int32 color_scheme { get; set; }
+
+    construct {
+        setup_user_path ();
+        setup_prefers_color_scheme ();
+    }
+
+    private void setup_user_path () {
+        try {
+            accounts_service = GLib.Bus.get_proxy_sync (
+                GLib.BusType.SYSTEM,
+               "org.freedesktop.Accounts",
+               "/org/freedesktop/Accounts"
+            );
+
+            user_path = accounts_service.find_user_by_name (GLib.Environment.get_user_name ());
+        } catch (Error e) {
+            critical (e.message);
+        }
+    }
+
+    private void setup_prefers_color_scheme () {
+        try {
+            pantheon_act = GLib.Bus.get_proxy_sync (
+                GLib.BusType.SYSTEM,
+                "org.freedesktop.Accounts",
+                user_path,
+                GLib.DBusProxyFlags.GET_INVALIDATED_PROPERTIES
+            );
+
+            color_scheme = pantheon_act.prefers_color_scheme;
+
+            ((GLib.DBusProxy) pantheon_act).g_properties_changed.connect ((changed, invalid) => {
+                var value = changed.lookup_value ("PrefersColorScheme", new VariantType ("i"));
+                if (value != null) {
+                    color_scheme = value.get_int32 ();
+                }
+            });
+        } catch (Error e) {
+            critical (e.message);
+        }
+    }
+}
+
+[DBus (name = "org.freedesktop.impl.portal.Settings")]
+public class SettingsDaemon.Settings : GLib.Object {
+    private GLib.DBusConnection connection;
+
+    public uint32 version {
+        get { return 1; }
+    }
+
+    public signal void setting_changed (string namespace, string key, GLib.Variant value);
+
+    private AccountsServiceMonitor monitor;
+
+    public Settings (GLib.DBusConnection connection) {
+        this.connection = connection;
+    }
+
+    construct {
+        monitor = new AccountsServiceMonitor ();
+        monitor.notify["color-scheme"].connect (() => {
+            var color_scheme = new Variant.uint32 (monitor.color_scheme);
+            setting_changed ("org.freedesktop.appearance", "color-scheme", get_color_scheme ());
+        });
+    }
+
+    private bool namespace_matches (string namespace, string[] patterns) {
+        foreach (var pattern in patterns) {
+            if (pattern[0] == '\0') {
+                return true;
+            }
+
+            if (pattern == namespace) {
+                return true;
+            }
+
+            int pattern_len = pattern.length;
+            if (pattern[pattern_len - 1] == '*' && Posix.strncmp (namespace, pattern, pattern_len - 1) == 0) {
```

Comment:
Do we want to depend on Posix here or reimplement this?

Reply:
I believe you can use `namespace.contains (pattern.slice (0, pattern_len - 1))`

Reply:
Ah indeed, though `s/contains/has_prefix/`